src/main/java/org/generationcp/middleware/dao/oms/CVTermDao.java

```diff
@@ -1381,4 +1418,331 @@ public CVTerm save(final String name, final String definition, final CvId cvId)
+		final String cropName) {
+
+		final SQLQuery sqlQuery = this.getSession().createSQLQuery(this.createVariableQuery(false));
+		this.appendGetVariablesScalar(sqlQuery);
+		final List<String> variableTypeNames =
+			variableTypes.stream().map(i -> VariableType.getById(i).getName()).collect(Collectors.toList());
+		sqlQuery.setParameterList(VARIABLE_TYPE_NAMES, variableTypeNames);
+		return this.convertToVariableDTO(this.getVariableQueryResult(pageSize, pageNumber, sqlQuery), cropName, false);
+
+	}
+
+	/**
+	 * Count the variables associated to a study filtered by variableTypes.
+	 *
+	 * @param datasetId
+	 * @param variableTypes
+	 * @return
+	 */
+	public long countVariablesByDatasetId(final Integer datasetId, final List<Integer> variableTypes) {
+		try {
+
+			final SQLQuery sqlQuery = this.getSession().createSQLQuery(this.createCountVariableQuery(true));
+			final List<String> variableTypeNames =
+				variableTypes.stream().map(i -> VariableType.getById(i).getName()).collect(Collectors.toList());
+			sqlQuery.setParameterList(VARIABLE_TYPE_NAMES, variableTypeNames);
+			sqlQuery.setParameter("datasetId", datasetId);
+			sqlQuery.setParameterList("variableTypes", variableTypes);
+
+			return ((BigInteger) sqlQuery.uniqueResult()).longValue();
+
+		} catch (final HibernateException e) {
+			LOG.error(e.getMessage(), e);
+			throw new MiddlewareQueryException("Error in countVariablesByStudy() query in CVTermDao: " + e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * Gets the list of variables associated to a study filtered by variableTypes.
+	 *
+	 * @param pageSize
+	 * @param pageNumber
+	 * @param studyId
+	 * @param variableTypes
+	 * @param cropName
+	 * @return
+	 */
+	public List<VariableDTO> getVariablesByDatasetId(final Integer pageSize, final Integer pageNumber, final Integer datasetId,
+		final List<Integer> variableTypes, final String cropName) {
+
+		final SQLQuery sqlQuery = this.getSession().createSQLQuery(this.createVariableQuery(true));
+		this.appendGetVariablesScalar(sqlQuery);
+
+		final List<String> variableTypeNames =
+			variableTypes.stream().map(i -> VariableType.getById(i).getName()).collect(Collectors.toList());
+		sqlQuery.setParameterList(VARIABLE_TYPE_NAMES, variableTypeNames);
+		sqlQuery.setParameter("datasetId", datasetId);
+		sqlQuery.setParameterList("variableTypes", variableTypes);
+
+		return this.convertToVariableDTO(this.getVariableQueryResult(pageSize, pageNumber, sqlQuery), cropName, true);
+
+	}
+
+	public List<Map<String, Object>> getVariableQueryResult(final Integer pageSize, final Integer pageNumber, final SQLQuery sqlQuery) {
+		if (pageNumber != null && pageSize != null) {
+			sqlQuery.setFirstResult(pageSize * (pageNumber - 1));
+			sqlQuery.setMaxResults(pageSize);
+		}
+		sqlQuery.setResultTransformer(AliasToEntityMapResultTransformer.INSTANCE);
+		return sqlQuery.list();
+	}
+
+	private String createCountVariableQuery(final boolean isFilterByStudyId) {
```

Comment:
You can make return type of `createCountVariableQuery` as `SQLQuery `and also replace the parameter `isFilterByStudyId `as `Optional<Integer> datasetId' and also pass in variableTypes as parameter.

You can then set common parameters within the method:
'final List<String> variableTypeNames =
			variableTypes.stream().map(i -> VariableType.getById(i).getName()).collect(Collectors.toList());
sqlQuery.setParameterList(VARIABLE_TYPE_NAMES, variableTypeNames);`
And also set the `variableTypes `and `datasetId` accordingly if datasetId is present.